<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#To-download-Enron-Data-from-Web" data-toc-modified-id="To-download-Enron-Data-from-Web-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>To download Enron Data from Web</a></span></li><li><span><a href="#Process-Emails" data-toc-modified-id="Process-Emails-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Process Emails</a></span></li><li><span><a href="#Create-Bert-Embeddings-for-Emails" data-toc-modified-id="Create-Bert-Embeddings-for-Emails-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Create Bert Embeddings for Emails</a></span></li><li><span><a href="#Nearest-Neighbors---Jaccard" data-toc-modified-id="Nearest-Neighbors---Jaccard-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Nearest Neighbors - Jaccard</a></span></li><li><span><a href="#Nearest-Neighbors---Cosine" data-toc-modified-id="Nearest-Neighbors---Cosine-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Nearest Neighbors - Cosine</a></span></li><li><span><a href="#Doc2Vec" data-toc-modified-id="Doc2Vec-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Doc2Vec</a></span></li><li><span><a href="#Precision-&amp;-Recall" data-toc-modified-id="Precision-&amp;-Recall-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Precision &amp; Recall</a></span></li></ul></div>

# To download Enron Data from Web

In [1]:
from get_enron import download_enron_data

In [2]:
download_enron_data(download_to_folder = os.path.join("input_data", "emails"))

File https://s3.amazonaws.com/freeeed.org/enron/results/enron021.zip - doesn't exist. Skipping this download
File https://s3.amazonaws.com/freeeed.org/enron/results/enron029.zip - doesn't exist. Skipping this download
Bad File Format https://s3.amazonaws.com/freeeed.org/enron/results/enron060.zip. Skipping this download
Bad File Format https://s3.amazonaws.com/freeeed.org/enron/results/enron061.zip. Skipping this download


# Process Emails

In [6]:
import os
from process_data import process_emails

In [7]:
process_emails(data_folder = os.path.join("input_data", "emails"), email_text_file = os.path.join("input_data", "enron_texts.csv"))

# Create Bert Embeddings for Emails

In [8]:
from document_representations_bert import get_doc_representations

In [9]:
get_doc_representations(input_email_text_file=os.path.join("input_data", "enron_texts.csv"),
                        output_embedding_file=os.path.join("input_data", "enron_embeddings.csv"))

10/19/2019 19:17:10 - INFO - pytorch_pretrained_bert.file_utils -   https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-vocab.txt not found in cache, downloading to /tmp/tmpuyad7ydd
100%|██████████| 231508/231508 [00:00<00:00, 6763826.49B/s]
10/19/2019 19:17:10 - INFO - pytorch_pretrained_bert.file_utils -   copying /tmp/tmpuyad7ydd to cache at /home/ubuntu/.pytorch_pretrained_bert/26bc1ad6c0ac742e9b52263248f6d0f00068293b33709fae12320c0e35ccfbbb.542ce4285a40d23a559526243235df47c5f75c197f04f37d1a0c124c32c9a084
10/19/2019 19:17:10 - INFO - pytorch_pretrained_bert.file_utils -   creating metadata file for /home/ubuntu/.pytorch_pretrained_bert/26bc1ad6c0ac742e9b52263248f6d0f00068293b33709fae12320c0e35ccfbbb.542ce4285a40d23a559526243235df47c5f75c197f04f37d1a0c124c32c9a084
10/19/2019 19:17:10 - INFO - pytorch_pretrained_bert.file_utils -   removing temp file /tmp/tmpuyad7ydd
10/19/2019 19:17:10 - INFO - pytorch_pretrained_bert.tokenization -   loading vocabulary file https:

10/19/2019 23:26:40 - INFO - root -   embedding chunk number: 74/124
10/19/2019 23:29:59 - INFO - root -   embedding chunk number: 75/124
10/19/2019 23:33:18 - INFO - root -   embedding chunk number: 76/124
10/19/2019 23:36:38 - INFO - root -   embedding chunk number: 77/124
10/19/2019 23:40:17 - INFO - root -   embedding chunk number: 78/124
10/19/2019 23:43:34 - INFO - root -   embedding chunk number: 79/124
10/19/2019 23:47:00 - INFO - root -   embedding chunk number: 80/124
10/19/2019 23:50:31 - INFO - root -   embedding chunk number: 81/124
10/19/2019 23:53:50 - INFO - root -   embedding chunk number: 82/124
10/19/2019 23:57:10 - INFO - root -   embedding chunk number: 83/124
10/20/2019 00:00:39 - INFO - root -   embedding chunk number: 84/124
10/20/2019 00:03:59 - INFO - root -   embedding chunk number: 85/124
10/20/2019 00:07:21 - INFO - root -   embedding chunk number: 86/124
10/20/2019 00:10:50 - INFO - root -   embedding chunk number: 87/124
10/20/2019 00:14:09 - INFO - root 

10/20/2019 06:05:39 - INFO - root -   embedding chunk number: 192/124
10/20/2019 06:09:02 - INFO - root -   embedding chunk number: 193/124
10/20/2019 06:12:22 - INFO - root -   embedding chunk number: 194/124
10/20/2019 06:15:42 - INFO - root -   embedding chunk number: 195/124
10/20/2019 06:19:02 - INFO - root -   embedding chunk number: 196/124
10/20/2019 06:22:22 - INFO - root -   embedding chunk number: 197/124
10/20/2019 06:25:58 - INFO - root -   embedding chunk number: 198/124
10/20/2019 06:29:19 - INFO - root -   embedding chunk number: 199/124
10/20/2019 06:32:38 - INFO - root -   embedding chunk number: 200/124
10/20/2019 06:35:59 - INFO - root -   embedding chunk number: 201/124
10/20/2019 06:39:16 - INFO - root -   embedding chunk number: 202/124
10/20/2019 06:42:34 - INFO - root -   embedding chunk number: 203/124
10/20/2019 06:45:59 - INFO - root -   embedding chunk number: 204/124
10/20/2019 06:49:21 - INFO - root -   embedding chunk number: 205/124
10/20/2019 06:53:14 

KeyboardInterrupt: 

Stopped above line after a while, as there are 1043264 records and 124 records per chunk, we would need to process 8414 times, and this would take more than a week to process, hence stopped. We still have good amount of data, i.e., 247*124 = 30628

# Nearest Neighbors - Jaccard

In [1]:
import os
from nn_similarity import train_nn, test_nn

In [2]:
# train model
train_nn(embeddings_file_path = os.path.join("input_data", "enron_embeddings.csv"), 
         distance_type="jaccard", 
         file_prefix="enron")

Model saved to: input_data/enron_jaccard_model.pkl


In [5]:
# test model on same train data
test_nn(embeddings_file_path = os.path.join("input_data", "enron_embeddings.csv"), 
        distance_type="jaccard", 
        file_prefix="enron")

# Nearest Neighbors - Cosine

In [2]:
import os
from nn_similarity import train_nn, test_nn

In [7]:
# train model
train_nn(embeddings_file_path = os.path.join("input_data", "enron_embeddings.csv"), 
         distance_type="cosine", 
         file_prefix="enron")

Model saved to: input_data/enron_cosine_model.pkl


In [3]:
# test model on same train data
test_nn(embeddings_file_path = os.path.join("input_data", "enron_embeddings.csv"), 
        distance_type="cosine",
       file_prefix="enron")

# Doc2Vec

In [4]:
import os
from doc2vec_similarity import train_doc2vec, test_doc2vec

In [2]:
train_doc2vec(model_file_name = os.path.join("input_data", "enron_doc2vec_model.bin"), 
              embeddings_file_path = os.path.join("input_data", "enron_embeddings.csv"))

Model saved to: input_data/doc2vec_model.bin


In [5]:
test_doc2vec(embeddings_file_path = os.path.join("input_data", "enron_embeddings.csv"), 
             model_file_name = os.path.join("input_data", "enron_doc2vec_model.bin"),
             results_path = os.path.join("output_data", "enron_doc2vec_testing.csv"))

Results saved to: input_data/doc2vec_model.bin


# Precision & Recall

In [8]:
import os
from precision_recall import p_r_f1_scores

In [10]:
p_r_f1_scores(jacc_results_path       = os.path.join("output_data", "enron_bert_testing_jaccard.csv"),
              cos_results_path        = os.path.join("output_data", "enron_bert_testing_cosine.csv"),
              doc2vec_results_path    = os.path.join("output_data", "enron_doc2vec_testing.csv"))


Model Name:  bert_jaccard
Precision:  bert_jaccard 1.0
Recall:  bert_jaccard 1.0
F1 score:  bert_jaccard 1.0

Model Name:  bert_cosine
Precision:  bert_cosine 1.0
Recall:  bert_cosine 1.0
F1 score:  bert_cosine 1.0

Model Name:  doc2vec
Precision:  doc2vec 1.0
Recall:  doc2vec 0.5570369406867846
F1 score:  doc2vec 0.7155089595255002
